libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,727 @@
+
+    // check if memory allocation failed
+    if (_path == nullptr || _prune.loops == nullptr || _simplify.stack == nullptr) {
+        log_action(SRTL_DEACTIVATED_INIT_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING, "SafeRTL deactivated: init failed");
+        free(_path);
+        free(_prune.loops);
+        free(_simplify.stack);
+        return;
+    }
+
+    _path_points_max = _points_max;
+
+    // create semaphore
+    _path_sem = hal.util->new_semaphore();
+
+    // when running the example sketch, we want the cleanup tasks to run when we tell them to, not in the background (so that they can be timed.)
+    if (!_example_mode){
+        // register background cleanup to run in IO thread
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_SafeRTL::run_background_cleanup, void));
+    }
+}
+
+// returns number of points on the path
+uint16_t AP_SafeRTL::get_num_points() const
+{
+    return _path_points_count;
+}
+
+// get next point on the path to home, returns true on success
+bool AP_SafeRTL::pop_point(Vector3f& point)
+{
+    // check we are active
+    if (!_active) {
+        return false;
+    }
+
+    // get semaphore
+    if (!_path_sem->take_nonblocking()) {
+        log_action(SRTL_POP_FAILED_NO_SEMAPHORE);
+        return false;
+    }
+
+    // check we have another point
+    if (_path_points_count == 0) {
+        _path_sem->give();
+        return false;
+    }
+
+    // return last point and remove from path
+    point = _path[--_path_points_count];
+
+    _path_sem->give();
+    return true;
+}
+
+// clear return path and set home location.  This should be called as part of the arming procedure
+void AP_SafeRTL::reset_path(bool position_ok)
+{
+    Vector3f current_pos;
+    position_ok &= _ahrs.get_relative_position_NED_origin(current_pos);
+    reset_path(position_ok, current_pos);
+}
+
+void AP_SafeRTL::reset_path(bool position_ok, const Vector3f& current_pos)
+{
+    if (_path == nullptr) {
+        return;
+    }
+
+    // clear path
+    _path_points_count = 0;
+
+    // reset simplification and pruning.  These functions access members that should normally only
+    // be touched by the background thread but it will not be running because active should be false
+    reset_simplification(0);
+    reset_pruning(0);
+
+    // de-activate if no position at take-off
+    if (!position_ok) {
+        deactivate(SRTL_DEACTIVATED_BAD_POSITION, "bad position");
+        return;
+    }
+
+    // save current position as first point in path
+    if (!add_point(current_pos)) {
+        deactivate(SRTL_ADD_FAILED_NO_SEMAPHORE, "failed to save initial point");
+        return;
+    }
+
+    // successfully added point and reset path
+    _last_good_position_ms = AP_HAL::millis();
+    _active = true;
+}
+
+// call this at 3hz (or higher) regardless of what mode the vehicle is in
+void AP_SafeRTL::update(bool position_ok, bool save_position)
+{
+    if (!_active || !position_ok || !save_position) {
```

Comment:
Checking `position_ok` here seems like a waste, since the general case is that it will be, we then and on to it, and check again in the next function call, why not only check it then?